cli/tools/lint/ast_buffer/swc.rs

```diff
@@ -0,0 +1,3018 @@
+      let async_pos = ctx.bool_field(AstProp::Async);
+      let gen_pos = ctx.bool_field(AstProp::Generator);
+      let id_pos = ctx.ref_field(AstProp::Id);
+      let type_params_pos = ctx.ref_field(AstProp::TypeParameters);
+      let return_pos = ctx.ref_field(AstProp::ReturnType);
+      let body_pos = ctx.ref_field(AstProp::Body);
+      let params_pos =
+        ctx.ref_vec_field(AstProp::Params, node.function.params.len());
+
+      let ident_id = serialize_ident(ctx, &node.ident, parent);
+      let type_param_id =
+        maybe_serialize_ts_type_param(ctx, &node.function.type_params, pos);
+      let return_type =
+        maybe_serialize_ts_type_ann(ctx, &node.function.return_type, pos);
+
+      let body = node
+        .function
+        .body
+        .as_ref()
+        .map(|body| serialize_stmt(ctx, &Stmt::Block(body.clone()), pos));
+
+      let params = node
+        .function
+        .params
+        .iter()
+        .map(|param| serialize_pat(ctx, &param.pat, pos))
+        .collect::<Vec<_>>();
+
+      ctx.write_bool(declare_pos, node.declare);
+      ctx.write_bool(async_pos, node.function.is_async);
+      ctx.write_bool(gen_pos, node.function.is_generator);
+      ctx.write_ref(id_pos, ident_id);
+      ctx.write_maybe_ref(type_params_pos, type_param_id);
+      ctx.write_maybe_ref(return_pos, return_type);
+      ctx.write_maybe_ref(body_pos, body);
+      ctx.write_refs(params_pos, params);
+
+      pos
+    }
+    Decl::Var(node) => {
+      let id = ctx.header(AstNode::VariableDeclaration, parent, &node.span, 3);
+      let declare_pos = ctx.bool_field(AstProp::Declare);
+      let kind_pos = ctx.str_field(AstProp::Kind);
+      let decls_pos =
+        ctx.ref_vec_field(AstProp::Declarations, node.decls.len());
+
+      let children = node
+        .decls
+        .iter()
+        .map(|decl| {
+          let child_id =
+            ctx.header(AstNode::VariableDeclarator, id, &decl.span, 2);
+          let id_pos = ctx.ref_field(AstProp::Id);
+          let init_pos = ctx.ref_field(AstProp::Init);
+
+          // FIXME: Definite?
+
+          let ident = serialize_pat(ctx, &decl.name, child_id);
+
+          let init = decl
+            .init
+            .as_ref()
+            .map(|init| serialize_expr(ctx, init.as_ref(), child_id));
+
+          ctx.write_ref(id_pos, ident);
+          ctx.write_maybe_ref(init_pos, init);
+
+          child_id
+        })
+        .collect::<Vec<_>>();
+
+      ctx.write_bool(declare_pos, node.declare);
+      ctx.write_str(
+        kind_pos,
+        match node.kind {
+          VarDeclKind::Var => "var",
+          VarDeclKind::Let => "let",
+          VarDeclKind::Const => "const",
+        },
+      );
+      ctx.write_refs(decls_pos, children);
+
+      id
+    }
+    Decl::Using(_) => {
+      todo!();
+    }
+    Decl::TsInterface(node) => {
+      let pos = ctx.header(AstNode::TSInterface, parent, &node.span, 0);
```

Comment:
Is this property count wrong? I feel like we need a way to express this in the code so that it doesn't need to be manually maintained and the property count can be automatically populated.

Reply:
Agree, those property counts are a bit brittle. I'll remove them once I'm back.